### PR TITLE
WIP : Python version of train scripts

### PR DESCRIPTION
**Pull request by vijayaditya**
This is a very preliminary draft of the python training script. This PR has been created for review of basic structure of the code. There might be several errors. 

The script exploits python multi-line string enclosed by """" """" to avoid escaping the " and '  characters in the queue.pl and run.pl commands being run.

### CODE REVIEW

File: egs/wsj/s5/steps/nnet3/train_rnn.py [hunk 1]
```diff
@@ -0,0 +1,1134 @@
...
+
+    if wait:
+        [stdout, stderr] = p.communicate()
+        if p.returncode is not 0:
+            raise Exception("There was an error while running the command {0}".format(command)
+        return stdout, stderr
+    else:
+        return p
+
+def GetSuccessfulModes(num_models, log_file_pattern):
```
Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/train_rnn.py]: i think you mean models, not modes

File: egs/wsj/s5/steps/nnet3/train_rnn.py [hunk 2]
```diff
@@ -0,0 +1,1134 @@
...
+
+def ForceSymlink(file1, file2):
+    try:
+        os.symlink(file1, file2)
+    except OSError, e:
+        if e.errno == errno.EEXIST:
+            os.remove(file2)
+            os.symlink(file1, file2)
+
+def ComputePresoftmaxPriorScale(dir, alidir, num_jobs, run_opts,
```
Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/train_rnn.py]: I'm not sure that we're using this prior-scale thing in the latest example scripts-- IIRC in the end we found it wasn't helpful (?)

File: egs/wsj/s5/steps/nnet3/train_rnn.py [hunk 3]
```diff
@@ -0,0 +1,1134 @@
...
+    for iter in range(num_iters):
+        if (args.exit_stage is not None) and (iter == args.exit_stage):
+            logger.info("Exiting early due to --exit-stage {0}".format(iter))
+            return
+        current_num_jobs = int(0.5 + args.num_jobs_initial + (args.num_jobs_final - args.num_jobs_initial) * iter / num_iters)
+        
+        if iter in realign_iters:
+            prev_egs_dir=cur_egs_dir
+            cur_egs_dir="{0}/egs_{1}".format(args.dir, "iter"+str(iter))
+            Realign(args.dir, iter, args.alidir, args.lang,
```
Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/train_rnn.py]: this should probably be inside a check for 'stage'.  If it isn't in the baseline script it would be a bug.

File: egs/wsj/s5/steps/nnet3/train_rnn.py [hunk 4]
```diff
@@ -0,0 +1,1134 @@
...
+        logger.warning("""
+    Without using a GPU this will be very slow.  nnet3 does not yet support multiple threads.""")
+        
+        run_opts.train_queue_opt = ""
+        run_opts.parallel_train_opts = "--use-gpu=no"
+        run_opts.combine_queue_opt = ""
+        run_opts.prior_gpu_opt = "--use-gpu=no"
+        run_opts.prior_queue_opt = ""
+    
+    if args.run_in == "cluster":
```
Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/train_rnn.py]: Vijay, this destroys all the power of the $cmd variable, just having two options like that, and it wouldn't be compatible with existing scripts.  It's not a rule that $cmd can only be one of two options-- may people use their own scripts.

Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/train_rnn.py]: OK. I will just change it to cmd, as in current script.

File: egs/wsj/s5/steps/nnet3/train_rnn.py [hunk 5]
```diff
@@ -0,0 +1,1132 @@
...
+    combine_queue_opt
+    prior_gpu_opt
+    prior_queue_opt
+    parallel_train_opts
+    
+    def __init__():
+        pass
+
+
+if __name__ == "__main__":
```
Comment (vdp) [inline egs/wsj/s5/steps/nnet3/train_rnn.py]: Just a couple, mostly stylistic/cosmetic, suggestions. I'm by no means a Python guru, so I don't claim that what I propose is necessarily better than the current code..

Alternatively you might use:

```
def main():
  <actual top-level code here>

if __name__ == '__main__':
  main()
```

That way the variables in the body of main() will be local. As it is now, *args* and *run_opts*, for example, are in fact visible in all functions. This is usually is not a problem in my experience, but once in a while may cause some head-scratching.

File: egs/wsj/s5/steps/nnet3/train_rnn.py [hunk 6]
```diff
@@ -0,0 +1,1132 @@
...
+    # we add compulsary arguments as named arguments for readability
+    parser = argparse.ArgumentParser(description="""
```
Comment (vdp) [inline egs/wsj/s5/steps/nnet3/train_rnn.py]: In my Python code I prefer to put the command line parsing code in a separate function and place it at the top, for easier reading. In fact as I think of this now, it might be better to place the main() code at the top as well(perhaps after option parsing), especially when the script is relatively long as in this case, because this is the logical place from which to start reading the code.
By the way, and I'm sure you know this already, you could add **formatter_class=argparse.ArgumentDefaultsHelpFormatter** as a parameter to ArgumentParser's constructor- will print out the default value for the options, when you call the script with "--help".

Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/train_rnn.py]: Thanks for the suggestion I will do this.

File: egs/wsj/s5/steps/nnet3/train_rnn.py (hunk 5 above)
Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/train_rnn.py]: Thanks a lot for pointing this out. I didn't know that definitions in __main__ were global.

**Comment by vijayaditya** [comment]
I have removed most of the bugs from the script. Training seems to run fine. I will be polishing the script further. But I think there will not be major changes and hence the script is ready for review.

I decided to create a kaldi_lib.py with training steps which are shared across feed-forward DNN and RNN training. Though this violates the initial goal of having a self-contained script it helps users identify steps which are unique to FF-DNN and RNN, as they will be coded in train_dnn.py and train_rnn.py.
I think this will help us better maintain the scripts in the future.

I have moved the make_configs.py script from train script to the local/nnet3/run_lstm.sh scripts, to ensure we can reuse the train_rnn.py for different RNN architectures. I will be doing the same for train_dnn.py too.

I am not too interested in changing decode.sh script as it already small and we will not gain a lot in terms of readability.

I am working on lstm chain script now, for which I will reuse a lot of steps from kaldi_lib.py

**Comment by danpovey** [comment]
maybe nnet3_train_lib.py would be a better name than kaldi_lib.py

**Comment by vijayaditya** [comment]
Yes that is better, modifying the name.

**Comment by vijayaditya** [comment]
Tested everything including the realignment pipeline. Realignment pipeline does not work due to issues raised in #495 . There are no more changes from my side, please review. 

Note : Experiment to reproduce results is running.

File: egs/wsj/s5/steps/nnet3/train_rnn.py [hunk 7]
```diff
@@ -0,0 +1,657 @@
...
+                        help="If the derivative averages are below this threshold we scale the parameter matrices with the shrink-value. It is less than 0.25 for sigmoid non-linearities.")
+
+    # RNN specific trainer options
+    parser.add_argument("--trainer.rnn.num-chunk-per-minibatch", type=int, dest='num_chunk_per_minibatch',
+                        default=100,
+                        help="Number of sequences to be processed in parallel every minibatch" )
+    parser.add_argument("--trainer.rnn.num-bptt-steps", type=int, dest='num_bptt_steps',
+                        default=None,
+                        help="The number of time steps to back-propagate from the last label in the chunk. By default it is same as the chunk-width." )
+    parser.add_argument("--trainer.rnn.shrink", type=float, dest='shrink',
```
Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/train_rnn.py]: just a note on shrinking, although it won't affect this pull request for now.
I am working in ~dpovey/kaldi-chain/ on a more elegant mechanism called 'self-repairing nonlinearities'.  It's an option in the nonlinearities (ReLU, sigmoid, tanh) where once the average value or derivative stats (as appropriate) get out of the 'good' range, it adds a small term to the derivative to try to force it back to a good range.  Hopefully this will work better than the shrinking mechanism- certainly it's more general.

File: egs/wsj/s5/steps/nnet3/train_rnn.py [hunk 8]
```diff
@@ -0,0 +1,657 @@
+#!/usr/bin/env python
+
```
Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/train_rnn.py]: There needs to be a copyright notice at the top of these files (just include your own name).

File: egs/wsj/s5/steps/nnet3/nnet3_train_lib.py [hunk 9]
```diff
@@ -0,0 +1,622 @@
...
+    print num_iters_combine
+    for iter in range(num_iters - num_iters_combine + 1, num_iters + 1):
+      model_file = '{0}/{1}.mdl'.format(dir, iter)
+      if not os.path.exists(model_file):
+          raise Exception('Model file {0} missing'.format(model_file))
+      raw_model_strings.append('"nnet3-am-copy --raw=true {0} -|"'.format(model_file))
+
+    combine_num_chunk_per_minibatch = int(1024.0/(chunk_width))
+    
+   # Below, we use --use-gpu=no to disable nnet3-combine-fast from using a GPU,
```
Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/nnet3_train_lib.py]: this comment is out of date, remove it.

Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/nnet3_train_lib.py]: ... also from the baseline bash script.

File: egs/wsj/s5/steps/nnet3/nnet3_train_lib.py [hunk 10]
```diff
@@ -0,0 +1,622 @@
...
+                 remove_egs = True):
+    try:
+        if remove_egs:
+            RemoveEgs(egs_dir)
+
+        for iter in range(num_iters):
+            print("deleting iter {0}".format(iter))
+            RemoveModel(nnet_dir, iter, num_iters, 1,
+                        preserve_model_interval)
+    except (IOError, OSError) as err:
```
Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/nnet3_train_lib.py]: i see no need for try-except here as there seem to be no expected errors-- or perhaps I'm missing something.  anyway it's just re-throwing the error.

Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/nnet3_train_lib.py]: I am catching the exception just to add some logging info, nothing more than that.

**Comment by danpovey** [comment]
How well tested is this?

**Comment by vijayaditya** [comment]
The experiment where I try to recreate the results of the train.sh script
is near  completion.

Vijay

**Comment by vijayaditya** [comment]
Tested end-to-end.

Added new reporting capability, where the training script sends periodic
update emails, if asked to, about the progress of the experiment or any error which
kills the experiment.

**Comment by vijayaditya** [comment]
any other comments/suggestions ?

**Comment by danpovey** [comment]
Regarding the periodic update emails-- bear in mind that not all clusters
will allow you to send email from the command line, and configuring email
is becoming increasingly hard due to spam-fighting efforts.  We don't want
to get into a situation where people say "why doesn't it work" and we have
to explain to them how to install postfix and work around local firewall
policies.  Perhaps wording it right will be enough?

Dan

**Comment by vijayaditya** [comment]
OK, I will add a disclaimer saying this is experimental and needs local
expertise.

Also, when there is an error in sending the email the logger just reports
inability to send the mail and continues with the training.

--Vijay

**Comment by vijayaditya** [comment]
added a disclaimer near the email option.

<pre><code>
+                        help=""" Email-id to report about the progress of the experiment.
 +                              NOTE: This is a very experimental feature. It assumes
 +                              the machine on which the script is being run can send
 +                              emails from command line via. mail program. The
 +                              Kaldi mailing list will not support this feature.
 +                              It might require local expertise to setup. """)
</code></pre>

**Comment by danpovey** [comment]
OK, but not sure that "experimental feature" is the right wording, maybe
remove that sentence.

**Comment by vijayaditya** [comment]
Done.

**Comment by danpovey** [comment]
ok but not ready to merge yet- will look again when I get a chance

**Comment by vijayaditya** [comment]
Could I get some comments on this PR. @tomkocse and @freewym  are close to completing their CNN and GRU recipes and it will be preferable if we ensure that there is no rework to later translate these recipes to use the python scripts.

File: egs/wsj/s5/steps/nnet3/nnet3_train_lib.py [hunk 11]
```diff
@@ -0,0 +1,618 @@
...
+
+def ForceSymlink(file1, file2):
+    try:
+        os.symlink(file1, file2)
+    except OSError, e:
+        if e.errno == errno.EEXIST:
+            os.remove(file2)
+            os.symlink(file1, file2)
+
+def ComputePresoftmaxPriorScale(dir, alidir, num_jobs, run_opts,
```
Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/nnet3_train_lib.py]: check if we're using this in any recipes- if not we could remove it.

File: egs/wsj/s5/steps/nnet3/train_rnn.py [hunk 12]
```diff
@@ -0,0 +1,703 @@
...
+    try:
+        Train(args, run_opts)
+    except Exception as e:
+        if args.email is not None:
+            message = "Training session for experiment {dir} died due to an error.".format(dir = args.dir)
+            sendMail(message, message, args.email)
+        traceback.print_exc()
+        raise e
+         
+def sendMail(message, subject, email_id):
```
Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/train_rnn.py]: is there a reason this is not SendMail?

**Comment by danpovey** [comment]
these issues are small so I'll merge this.

**Comment by danpovey** [comment]
oops, I see there are conflicts, I can't merge.

File: egs/wsj/s5/steps/nnet3/train_rnn.py (hunk 12 above)
Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/train_rnn.py]: Changing..

File: egs/wsj/s5/steps/nnet3/nnet3_train_lib.py (hunk 11 above)
Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/nnet3_train_lib.py]: We are using it in the TDNN recipes. This file is shared by both train_dnn.py and train_rnn.py

**Comment by vijayaditya** [comment]
resolving the conflicts.

**Comment by vijayaditya** [comment]
resolved all conflicts.
